db/persistent_stats_history.cc

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+// Copyright (c) 2011 The LevelDB Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file. See the AUTHORS file for names of contributors.
+
+#include "db/persistent_stats_history.h"
+#include <string>
+#include <utility>
+#include "db/db_impl.h"
+#include "util/string_util.h"
+
+namespace rocksdb {
+const int kNowMicrosStringLength = 16;
+
+PersistentStatsHistoryIterator::~PersistentStatsHistoryIterator() {}
+
+bool PersistentStatsHistoryIterator::Valid() const { return valid_; }
+
+Status PersistentStatsHistoryIterator::status() const { return status_; }
+
+void PersistentStatsHistoryIterator::Next() {
+  // increment start_time by 1 to avoid infinite loop
+  AdvanceIteratorByTime(GetStatsTime() + 1, end_time_);
+}
+
+uint64_t PersistentStatsHistoryIterator::GetStatsTime() const { return time_; }
+
+const std::map<std::string, uint64_t>&
+PersistentStatsHistoryIterator::GetStatsMap() const {
+  return stats_map_;
+}
+
+std::pair<uint64_t, std::string> parseKey(std::string key,
+                                          uint64_t start_time) {
+  std::pair<uint64_t, std::string> result;
+  std::string::size_type pos = key.find("#");
+  // TODO(Zhongyi): add counters to track parse failures?
+  if (pos == std::string::npos) {
+    result.first = 0;
+    result.second = "";
+  } else {
+    uint64_t parsed_time =
+        static_cast<uint64_t>(std::stoull(key.substr(0, pos)));
+    // skip entries with timestamp smaller than start_time
+    if (parsed_time < start_time) {
+      result.first = 0;
+      result.second = "";
+    } else {
+      result.first = parsed_time;
+      std::string key_resize = key.substr(pos + 1);
+      key_resize.erase(std::find(key_resize.begin(), key_resize.end(), '\0'), key_resize.end());
```

Comment:
What is it for?

Reply:
it's a fail-safe to handle the case that key was improperly constructed with multiple '\0' at the end or there are garbage after the first '\0'.

Reply:
this line will be removed